text/002x-succinct-atomic-swaps.md

```diff
@@ -0,0 +1,170 @@
+
+The reference implementation uses Bitcoin as the other chain in the swap, but potentially any other chain using Secp256k1 keys and multisignatures can work with the protocol.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Funds on the non-Grin chain need to be locked in a multisignature transaction that requires knowledge of both "atomic nonces" to be spent. A refund transaction is also needed to recover funds in case one of the parties becomes non-responsive, or the protocol needs to be aborted for another reason.
+
+## Atomic nonces
+[atomic-nonces]: #atomic-nonces
+
+In the case of Grin, atomic nonces are just Secp256k1 private keys used to unlock funds on the other chain. Let's say Alice knows atomic nonce `aN`, and Bob knows atomic nonce `bN`.
+
+In the refund transaction, Alice (with locked Grin funds) will reveal her nonce `aN` to allow Bob to recover his locked funds on the other chain.
+
+In the main transaction, Bob and Alice create a multisignature Grin transaction which reveals Bob's nonce `bN` to Alice. Alice can then recover the funds locked on the other chain.
+
+Atomic IDs are identifiers used to derive atomic nonces using the keychain. Atomic IDs are formatted as follows:
+
+```
+0x4 | 'mwatomic' | 0x0 0x0 0x0 0x0 | 32-bit ID
+```
+
+The 32-bit ID can be expanded to 64-bits if necessary, and doesn't need to be random (it can be a simple counter). However, an ID can only be used once.
+
+Currently, output blinding factors and signature nonces use the key ID of the inputs and outputs, which is derived from the current child key path, or next available key path.
+
+Atomic swaps still use the default methods for deriving blinding factors and signature nonces. The atomic IDs are domain separated to help ensure they don't collide with normal blinding factors and signature nonces.
+
+Atomic nonces must only be used **once**. Reuse potentially results in loss of funds, if the other party has the previously used nonce.
+
+Without automated defenses, a malicious party could trick a user into deriving the same nonce (by supplying a duplicate atomic ID). The malicious user could then recover funds on the other chain without completing the atomic swap protocol.
+
+Regular blinding factors and signature nonces use a strictly increasing counter to prevent deriving the same factors more than once. However, this technique alone won't work for atomic nonces. The atomic ID is shared between parties, meaning a malicious attacker could attempt to trick a user into reusing an atomic ID.
```

Comment:
How can a user be tricked into reusing an atomic ID, when they are generated by an ever increasing counter?

Reply:
It depends. Currently, the atomic ID is passed across the refund and main transaction to coordinate the swap. I'm not sure how exactly to do the coordination without it.

Generating the atomic nonce using only local data could work, I just need to think about it more. Removing the need for an atomic ID would also allow for dropping the filter, though I think the domain separation with the prefix is still a good idea.

Using the prefix helps the user to not use atomic nonces as normal blinding factors and/or signature nonces on a wallet restore. Normally, reusing keys isn't that big of a problem. Reusing atomic nonces is a serious security risk.

Reply:
Another problem I am running into with just using the counter:

- Alice begins an atomic swap with Bob, `atomic_id = 0`
- Alice calls `next_atomic_id`, gets ID ending in `0`, `atomic_id = 1`
- Alice begins another atomic swap with Charlie before ending the one with Bob, `atomic_id = 1`

How should Alice recover the atomic id with Bob to finish the swap with him?

If we use the `output_id`, similar to how the `parent_key_id` is used, to store the `atomic_id` then there will only be one `atomic_id` stored per output. However, there would also need to be a global (per-`keychain_mask`) `atomic_id` stored to avoid reuse. Does this sound reasonable?

Reply:
I switched to deriving the atomic nonces entirely locally, using the counter. The atomic nonce's are recoverable from storage using the slate ID associated with the swap.